Mathlib/RingTheory/MvPowerSeries/Substitution.lean

```diff
@@ -0,0 +1,471 @@
+    (f : MvPowerSeries σ R) :
+    (substAlgHom hb) (substAlgHom ha f) = substAlgHom (ha.comp hb) f :=
+  DFunLike.congr_fun (substAlgHom_comp_substAlgHom ha hb) f
+
+theorem subst_comp_subst (ha : HasSubst a) (hb : HasSubst b) :
+    (subst b) ∘ (subst a) = subst (R := R) (fun s ↦ subst b (a s)) := by
+  simpa only [funext_iff, coe_substAlgHom, DFunLike.ext_iff,
+    AlgHom.coe_comp, AlgHom.coe_restrictScalars', Function.comp_apply]
+    using substAlgHom_comp_substAlgHom (R := R) ha hb
+
+theorem subst_comp_subst_apply (ha : HasSubst a) (hb : HasSubst b) (f : MvPowerSeries σ R) :
+    subst b (subst a f) = subst (fun s ↦ subst b (a s)) f :=
+  congr_fun (subst_comp_subst (R := R) ha hb) f
+
+section rescale
+
+/-- Rescale multivariate power series -/
+noncomputable def rescale (a : σ → R) (f : MvPowerSeries σ R) :
+    MvPowerSeries σ R :=
+  subst (a • X) f
+
+theorem rescale_eq_subst (a : σ → R) (f : MvPowerSeries σ R) :
+    rescale a f = subst (a • X) f := rfl
+
+variable (R) in
+theorem hasSubst_rescale (a : σ → R) :
+    HasSubst ((a • X) : σ → MvPowerSeries σ R) := by
+  convert hasSubst_mul (fun s ↦ algebraMap R (MvPowerSeries σ R) (a s))
+    hasSubst_X using 1
+  rw [funext_iff]
+  intro s
+  simp only [Pi.smul_apply', Pi.mul_apply]
+  rw [algebra_compatible_smul (MvPowerSeries σ R), smul_eq_mul]
+
+variable (R) in
+/-- Rescale multivariate power series, as an `AlgHom` -/
+noncomputable def rescale_algHom (a : σ → R) :
+    MvPowerSeries σ R →ₐ[R] MvPowerSeries σ  R :=
+  substAlgHom (hasSubst_rescale R a)
+
+theorem coe_rescale_algHom (a : σ → R) :
+    ⇑(rescale_algHom R a) = rescale a :=
+  coe_substAlgHom (hasSubst_rescale R a)
+
+theorem rescale_algHom_comp (a b : σ → R) :
+    (rescale_algHom R a).comp (rescale_algHom R b) = rescale_algHom R (a * b) := by
+  rw [AlgHom.ext_iff]
+  intro f
+  simp only [AlgHom.coe_comp, Function.comp_apply, rescale_algHom]
+  rw [substAlgHom_comp_substAlgHom_apply]
+  congr
+  rw [funext_iff]
+  intro s
+  simp only [Pi.smul_apply', Pi.mul_apply]
+  rw [AlgHom.map_smul_of_tower]
+  rw [← MvPolynomial.coe_X, substAlgHom_coe, MvPolynomial.aeval_X, MvPolynomial.coe_X]
+  simp only [Pi.smul_apply', algebraMap_smul]
+  rw [← mul_smul, mul_comm]
+
+theorem rescale_rescale_apply (a b : σ → R) (f : MvPowerSeries σ R) :
+    (f.rescale b).rescale a = f.rescale (a * b) := by
+  simp only [← coe_rescale_algHom, ← AlgHom.comp_apply, rescale_algHom_comp]
+
+theorem coeff_rescale (r : σ → R) (f : MvPowerSeries σ R) (d : σ →₀ ℕ) :
+    coeff R d (rescale r f) = (d.prod fun s n ↦ r s ^ n) • coeff R d f := by
+  unfold rescale
+  rw [coeff_subst (hasSubst_rescale R _)]
+  simp only [Pi.smul_apply', smul_eq_mul, prod_smul_X_eq_smul_monomial_one]
+  simp only [LinearMap.map_smul_of_tower, Algebra.mul_smul_comm]
+  rw [finsum_eq_single _ d]
+  · simp only [coeff_monomial_same, mul_one, smul_eq_mul]
+  · intro e he
+    simp only [coeff_monomial_ne he.symm, mul_zero, smul_zero]
```

Comment:
```suggestion
  unfold rescale
  rw [coeff_subst (hasSubst_rescale R _)]
  simp only [Pi.smul_apply', smul_eq_mul, prod_smul_X_eq_smul_monomial_one]
  simp only [LinearMap.map_smul_of_tower, Algebra.mul_smul_comm]
  rw [finsum_eq_single _ d]
  · simp
  · intro e he
    simp [coeff_monomial_ne he.symm]
```
Also, can you create a `rfl` lemma to rewrite instead of unfolding `rescale`?

Reply:
it existed already!